src/Bicep.Core.UnitTests/TypeSystem/Az/AzResourceTypeProviderTests.cs

```diff
@@ -29,31 +31,54 @@ public class AzResourceTypeProviderTests
-        [DataRow(ResourceTypeGenerationFlags.NestedResource)]
-        [DataRow(ResourceTypeGenerationFlags.ExistingResource | ResourceTypeGenerationFlags.HasParentDefined)]
-        public void AzResourceTypeProvider_can_deserialize_all_types_without_throwing(ResourceTypeGenerationFlags flags)
+        private static NamespaceType GetAzNamespaceType()
         {
-            var typeProvider = TestTypeHelper.CreateWithAzTypes();
-            var namespaceType = typeProvider.TryGetNamespace("az", "az", ResourceScope.ResourceGroup)!;
+            var nsProvider = new DefaultNamespaceProvider(new AzResourceTypeLoader(), BicepTestConstants.Features);
 
-            var resourceTypeProvider = namespaceType.ResourceTypeProvider;
-            var availableTypes = resourceTypeProvider.GetAvailableTypes();
+            return nsProvider.TryGetNamespace("az", "az", ResourceScope.ResourceGroup)!;
+        }
 
-            // sanity check - we know there should be a lot of types available
-            var expectedTypeCount = 3000;
-            availableTypes.Should().HaveCountGreaterThan(expectedTypeCount);
+        private static IEnumerable<object[]> GetDeserializeTestData()
+        {
+            var flagPermutationsToTest = new [] {
+                ResourceTypeGenerationFlags.None,
+                ResourceTypeGenerationFlags.ExistingResource,
+                ResourceTypeGenerationFlags.HasParentDefined,
+                ResourceTypeGenerationFlags.NestedResource,
+                ResourceTypeGenerationFlags.ExistingResource | ResourceTypeGenerationFlags.HasParentDefined,
+            };
+
+            foreach (var providerGrouping in GetAzNamespaceType().ResourceTypeProvider.GetAvailableTypes().GroupBy(x => x.TypeSegments[0])) {
+                foreach (var apiVersionGrouping in providerGrouping.GroupBy(x => x.ApiVersion)) {
+                    foreach (var flags in flagPermutationsToTest) {
+                        var providerName = providerGrouping.Key;
+                        var apiVersion = apiVersionGrouping.Key!;
+                        var resourceTypes = apiVersionGrouping.Select(x => x.FormatName()).ToList();
+
+                        yield return new object[] { providerName, apiVersion, flags, resourceTypes };
+                    }
+                }
+            }
+        }
+
+        public static string GetDeserializeTestDisplayName(MethodInfo info, object[] values)
+            => $"{info.Name} ({string.Join(',', new[] { values[0], values[1], values[2] }.Select(x => x.ToString()))})";
 
-            foreach (var availableType in availableTypes)
+        [DataTestMethod]
+        [DynamicData(nameof(GetDeserializeTestData), DynamicDataSourceType.Method, DynamicDataDisplayName = nameof(GetDeserializeTestDisplayName))]
+        public void AzResourceTypeProvider_can_deserialize_all_types_without_throwing(string providerName, string apiVersion, ResourceTypeGenerationFlags flags, IReadOnlyList<string> resourceTypes)
```

Comment:
>AzResourceTypeProvider_can_deserialize_all_types_without_throwing [](http://example.com/codeflow?start=20&length=65)

Btw, this is now the slowest test in VS on my machine. I'm thinking of making each type a data row, so they can run in parallel.

Reply:
It should already be heavily parallelizable with this change; even more parallelism may be counterproductive - I suspect the slowdown is probably because there's less cache utilization (or also just having to do more initialization as there are now many more data rows). Also note that this test is very CPU-bound. The reason I picked providers & apiversions as params (rather than individual types) is because it aligns with how the embedded JSON files are stored in the types DLL.

One option to speed it up may be to remove `flags` as a test parameter, and loop through each flag within the test itself - that way we'd be able to re-use the data [in this cache](https://github.com/Azure/bicep/blob/a9ff1f54ca773c16eb4b4471ce452fca795d2b68/src/Bicep.Core/TypeSystem/Az/AzResourceTypeFactory.cs#L14) which is expensive to re-compute.

Over time, this test is going to get worse and worse, as we constantly add new types and new data to the types library without removing anything - not sure if there's anything we can really do about that, aside from simply deciding to skip things for testing. The work to move types to the registry implies we'll have to change our strategy at some point in the future (e.g. presumably run this test in the types repo rather than here).